MultiTenant implementation

chadbengen opened the pull request:
I'm submitting a second attempt at multi-tenancy.  This version of it is significantly less obtrusive and more customizable.  I threw together a readme in the Skoruba.MultiTenant project.

Essentially i'm proposing a Skoruba abstraction that allows any implementation of tenant resolution, with a working implementation.

chadbengen commented:
@skoruba I found a bug and had to squash and rebase the PR.

skoruba commented:
Thanks, the tests failed, can you check it?

chadbengen commented:
@skoruba, the tests pass, but more thought needs to be put into how to do them properly.  Currently the tests pass by overriding the tenant configuration to be a single tenant configuration through the use of virtual voids in the setup.   I'm going to work on adding integration tests that do the same but for multitenant with in memory store.

But it should be well enough for a review and discussion.

skoruba commented:
@chadbengen - Thanks for your PR, I will review these changes with @aiscrim.

chadbengen commented:
I updated the tests so they run both the single tenant and the multi tenant.  However i found that the HttpClient in the BaseClassFixture does not clear cookies.  For instance, debugging the UserIsAbleToLogin test, i put a break point in the GET Account/Login method and saw that the user is authenticated, which i assume is a result of the registration.

In the SetupClient extension i added the option: HandleCookies = false.  This allowed tests to pass locally, but it failed in your appveyor.  I installed appveyor to my github fork and the tests pass. 

I'm assuming there's a race condition affecting the HttpClient.  I'm not sure how best to proceed, so i'm going to leave it be for the moment.  Let me know if you want me to disable the new multitenant tests.

skoruba commented:
Thanks, I will check it. 👍🏼

aiscrim commented on src/Skoruba.MultiTenant.Finbuckle/Configuration/RequestParam.cs:
```diff
@@ -0,0 +1,18 @@
+        public string Name { get; set; }
+
+        /// <summary>
+        /// 1 = Code, 2 = Identifier, 3 = Id
```
Can you use an enum here?

chadbengen replied:
Yes.  Done.

chadbengen commented on Skoruba.IdentityServer4.Admin.sln.startup.json:
```diff
@@ -0,0 +1,63 @@
+    /*  Create an item in the dropdown list for each project in the solution?  */
+    "ListAllProjects": true,
+
+    /*
```
That's an artifact of my dev environment and should not be there, so i removed it.

chadbengen commented on src/Skoruba.DbMigrator/Extensions/DbContextExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+﻿using Microsoft.EntityFrameworkCore;
```
Removed

chadbengen commented on src/Skoruba.IdentityServer4.Admin/Program.cs:
```diff
@@ -33,12 +34,11 @@ public static async Task Main(string[] args)
-                        .EnsureSeedData<IdentityServerConfigurationDbContext, AdminIdentityDbContext,
-                            IdentityServerPersistedGrantDbContext, AdminLogDbContext, AdminAuditLogDbContext,
-                            UserIdentity, UserIdentityRole>(host);
+                    //await DbMigrationHelpers
```
I didn't want to remove this code unless DbMigrator was agreed upon as a replacement. If you think that's the case then i will remove it.

softwareguy74 replied:
I think clients and resources per tenant absolutely makes sense.   Can we add that?

chadbengen commented:
@aiscrim Regarding the Clients and ApiResources, i have not considered any changes.  I dont see the advantage, personally.

aiscrim commented:
@chadbengen so your assumption is that all the tenants have access to all the applications integrated with the Identity Server? It's perfectly legit, I just wanted to make sure it's a decision and not just something being overlooked.
What's the use case you've in your mind for multi tenancy?
I was thinking of two different companies that externalize their identity provider to a third party hosting the Identity Server. In this case each of the companies would have different applications integrated with the same Identity Server, so the list of Clients and APIs would be different for the two tenants, like for Users and Roles.

chadbengen commented:
Yeah, that's not a use case i considered.  In my case the identity server is specific to my company's applications, which happen to be multi-tenant.  I'm not able to wrap around that use case right now, but i agree that it should be considered.

chadbengen commented:
@aiscrim I am adding new commits to a different branch, feature/multitenant-2-dev.  I dont want to introduce changes that make reviewing more challenging.  

**Changes:**
- Removed the ClaimsPrincipalFactory
- Implemented User/Role validators and simplified the stores.  This one introduced a lot of file changes.
- Added a MultiTenantException class

**Todo:**
- Add a tenant page for adding/editing tenants,

**Questions:**
- Should the MultiTenantEnabled flag be a configuration value that is injected rather than set as static? 
- Can the DbMigrator completely replace the current migration strategy, or should they co-exist?

**Considerations**
- Should there be an authorization policy allowing certain roles to be able to create/edit users and roles within any tenant?
- What would be necessary to allow tenant specific clients and api resources?  And can it be added later or should it be at the same time?

aiscrim commented:
@chadbengen thanks a lot! 
- MultiTenantEnabled flag: yes, I think that hard coding it statically it's not the best option. My question for you is: what happens if this parameter changes from true to false at some time? (sorry, I haven't looked at the whole commit yet)
- My opinion is that we don't duplicate functionality, so if we now have a DbMigrator and it's accepted by @skoruba , I don't see the utility of launching the migration from the Identity Server itself. I don't have a strong opinion on this though.
- Having a "super admin" role who's allowed to administer resources belonging to any tenants would certainly be an interesting feature. It can be postponed to a "phase 2", together with...
- ... tenant specific clients and api resources. They can be added later for sure. We could also give the possibility to specify in the configuration which resource types will be multi-tenant and which not...

chadbengen commented:
@aiscrim 

I went ahead and removed the the static MultiTenantEnabled flag.  This resulted in a pretty significant refactoring, but in the end has made it substantially easier to configure.  Essentially, just set the appsetting value to true or false and that's it.  

It's all in the feature/multitenant-2-dev branch.  Let me know if you're okay with me updating the PR branch.

I updated the readme so that it defines the use case and how the Finbuckle implementation solves the use case, as well as how the other services help.

My eyes are blurry.

aiscrim commented:
Excellent job @chadbengen. I second you move on with the merging of the multitenant-2-dev branch into this PR, it wouldn't make much sense for me to keep reviewing a version that has been heavily refactored.

aiscrim commented on src/Skoruba.MultiTenant.Finbuckle/Extensions/TenantInfoExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace Skoruba.MultiTenant.Finbuckle.Extensions
+{
+    public static class TenantInfoExtensions
```
I could find no code implementing the logic for the IsActive parameter. Please complete the functionality or remove the parameter.

chadbengen replied:
This will be implemented with the tenant manage page

chadbengen commented on src/Skoruba.IdentityServer4.STS.Identity/Controllers/AccountController.cs:
```diff
@@ -576,6 +581,16 @@ public async Task<IActionResult> Register(RegisterViewModel model, string return
 
+            if (_skorubaTenantContext.MultiTenantEnabled)
+            {
+                if (_skorubaTenantContext.TenantResolutionRequired && !_skorubaTenantContext.TenantResolutionRequired)
```
fixed

chadbengen commented on src/Skoruba.MultiTenant.Identity/Validators/UserRequiresTenantIdValidator.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            if (_skorubaTenantContext.TenantResolutionRequired && !_skorubaTenantContext.TenantResolved)
+            {
+                throw MultiTenantException.MissingTenant;
```
changed to modelstate error.

chadbengen commented on src/Skoruba.IdentityServer4.STS.Identity/Configuration/Test/StartupTestMultiTenant.cs:
```diff
@@ -0,0 +1,67 @@
+﻿using Microsoft.AspNetCore.Builder;
```
The file was originally located here and was named StartupTest.  The tests required a multi-tenant version of the file and so i created and renamed them to be more inline with their function.

chadbengen commented on src/Skoruba.IdentityServer4.STS.Identity/Configuration/Test/StartupTestSingleTenant.cs:
```diff
@@ -0,0 +1,42 @@
+﻿using Microsoft.AspNetCore.Builder;
```
The file was originally located here and was named StartupTest.  The tests required a multi-tenant version of the file and so i created and renamed them to be more inline with their function.

chadbengen commented on src/Skoruba.IdentityServer4.Admin/Configuration/Test/StartupTestSingleTenant.cs:
```diff
@@ -9,9 +9,9 @@
 
 namespace Skoruba.IdentityServer4.Admin.Configuration.Test
```
The file was originally located here and was named StartupTest.  The tests required a multi-tenant version of the file and so i created and renamed them to be more inline with their function.

chadbengen commented on src/Skoruba.IdentityServer4.Admin/Configuration/Test/StartupTestMultiTenant.cs:
```diff
@@ -0,0 +1,80 @@
+using Skoruba.MultiTenant.Stores;
+using System;
+
+namespace Skoruba.IdentityServer4.Admin.Configuration.Test
```
The file was originally located here and was named StartupTest.  The tests required a multi-tenant version of the file and so i created and renamed them to be more inline with their function.

chadbengen commented on src/Skoruba.MultiTenant.Identity/Validators/RoleRequiresTenantIdValidator.cs:
```diff
@@ -0,0 +1,35 @@
+        {
+            if (_skorubaTenantContext.TenantResolutionRequired && !_skorubaTenantContext.TenantResolved)
+            {
+                throw MultiTenantException.MissingTenant;
```
The reason i'm throwing an exception is because **it should not happen**.  If it does not resolve and is required then there is something wrong with the multi tenant implementation.  I worry that a graceful exit might hide a larger problem.

aiscrim replied:
Agreed.

chadbengen commented on src/Skoruba.MultiTenant.Identity/Validators/UserRequiresTwoFactorAuthenticationValidator.cs:
```diff
@@ -0,0 +1,60 @@
+
+                if (_skorubaTenantContext.TenantResolutionRequired && !_skorubaTenantContext.TenantResolved)
+                {
+                    throw MultiTenantException.MissingTenant;
```
The reason i'm throwing an exception is because **it should not happen**.  If it does not resolve and is required then there is something wrong with the multi tenant implementation.  I worry that a graceful exit might hide a larger problem.

aiscrim replied:
Agreed.

chadbengen commented on src/Skoruba.MultiTenant.Identity/Validators/UserRequiresTwoFactorAuthenticationValidator.cs:
```diff
@@ -0,0 +1,60 @@
+
+#if DEBUG
+                // Ignore seeded email
+                if (identityUser.Email.EndsWith("@skoruba.com"))
```
I removed it.  Its better to just remove the RequiresToFactorAuthentication item from the tenant seed data.

chadbengen commented on src/Skoruba.MultiTenant/MultiTenantException.cs:
```diff
@@ -0,0 +1,15 @@
+    /// </summary>
+    public class MultiTenantException : Exception
+    {
+        public static MultiTenantException MissingTenant => new MultiTenantException(MultiTenantConstants.MissingTenantExceptionMessage);
```
Agreed.  There's several areas that should be localized besides this.  I'm not strong with localization, so i'm going to wait until everything else in in the PR is cleaned up.

chadbengen commented on src/Skoruba.MultiTenant.EFCacheStore/EFCoreCacheStore.cs:
```diff
@@ -0,0 +1,123 @@
+        private const string _cacheKey = "Skoruba.MultiTenant.EfCoreCacheStore";
+
+        private readonly MemoryCacheEntryOptions _cacheOptions = new MemoryCacheEntryOptions()
+            .SetSlidingExpiration(TimeSpan.FromHours(48));
```
done.

chadbengen commented on src/Skoruba.IdentityServer4.STS.Identity/Views/Shared/_Layout.cshtml:
```diff
@@ -46,6 +49,9 @@
         @if (User.Identity.IsAuthenticated && !string.IsNullOrEmpty(name))
         {
+            <!--Tenant name-->
+            <p>@SkorubaTenantContext?.Tenant?.Name</p>
```
done.

chadbengen commented on src/Skoruba.IdentityServer4.Admin/Views/Shared/_Layout.cshtml:
```diff
@@ -29,6 +32,9 @@
         <h3 class="menu-logo my-0 mr-md-auto font-weight-normal"><a class="logo" asp-area="" asp-controller="Home" asp-action="Index">@RootConfiguration.AdminConfiguration.PageTitle</a></h3>
 
+        <!--Tenant name-->
+        <p>@SkorubaTenantContext.Tenant?.Name</p>
```
done

chadbengen commented on src/Skoruba.IdentityServer4.Admin/Controllers/IdentityController.cs:
```diff
@@ -77,12 +82,22 @@ public async Task<IActionResult> Role(TRoleDtoKey id)
+
+                if (_skorubaTenantContext.MultiTenantEnabled)
+                {
+                    // TODO: Is this business logic and should a new role be retrieved from a service, or is that over engineering?
```
I put it on the SkorubaTenantContext.  Seems like a reusable and reasonable responsibility.  Let me know if you dont agree.

chadbengen commented on src/Skoruba.MultiTenant/Configuration/MultiTenantConstants.cs:
```diff
@@ -0,0 +1,13 @@
+        public const string RequiresTwoFactorAuthentication = nameof(RequiresTwoFactorAuthentication);
+
+        // Optional tenant item key determining if the tenant is active or not
+        public const string IsActive = nameof(IsActive);
```
This is being refactored in another commit.

chadbengen commented on src/Skoruba.MultiTenant/Configuration/MultiTenantConstants.cs:
```diff
@@ -0,0 +1,13 @@
+        public const string MissingTenantExceptionMessage = "A tenant is required.";
+
+        // Optional tenant item key determining whether the tenant requires 2fa or not
+        public const string RequiresTwoFactorAuthentication = nameof(RequiresTwoFactorAuthentication);
```
This is being refactored in another commit.

chadbengen commented:
- Project Skoruba.MultiTenant.EntityFrameworkCore is removed.

- I'm okay if the default strategy changes to something else.  My desire is to see something that allows any strategy to be implemented, with the fewest changes necessary.  

- I appreciate the vote for the DbMigrator.  I added some enhancements to it so that it auto finds MigrateAndSeed objects so long as it has a reference to the project.  Although, it could probably be modified to work even without a reference.

The commits i made also introduced some other small changes.  For instance, i combined the two startup methods for middleware into just one.  

Here's a question that i'd like to get you and @skoruba  to weigh in on.  

One of my goals with this PR was to minimize the number of changes to existing projects, but implementing the controllers, services, views, etc. will require two or three additional projects to implement cleanly.  So my question is if it is better to put the services, mappers, etc. in with the existing Skoruba projects or should i create a few more MultiTenant projects and keep the multi tenant stuff separate?

The advantage of keeping it all separate is that multi tenant services could remain a second class citizen and only be pulled in if desired (other than the abstraction).  This would then also beg the question if the tests should be split out.

chadbengen commented:
Rebased onto dev branch

abodehq commented:
guys great work, when this feature will be published , also i want to check if each tenant will have its own users and roles ?? we want to create multi tenant ecommerce app which mean if user1 register for tenant1 he can also register for tenant2 with different password or info , is this will be supported in this version ??

please let us know asap.thaaaaaaaaaanks.

skoruba commented:
The multitanancy is currently out of scope of this project, so I am closing this PR.
Thank you for your effort on this PR - I really appreciate your work, maybe in the future I will take a look at this code and I will use it.
